### PR TITLE
Optimize handling of wide datasets and Pandas indexes 

### PR DESCRIPTION
**Pull request by maximlt**
By avoiding calling `redim` on each element of an `NdOverlay` which meant that every element had its own copy of the data

---

Edit: This PR bundles a few optimizations:

- https://github.com/holoviz/hvplot/pull/1350/commits/1e61065d041db889dbfed2634e5918e07295dab2: as described above, `.redim` is no longer called the elements of an `NdOverlay` so they can share the same data. This is demonstrated with the example below, the `data` wrapped by each element of the `NdOverlay` is now the same object.

```python
import hvplot.pandas  # noqa
from bokeh.sampledata.degrees import data as deg

plot = deg.hvplot.line(x='Year', y=['Art and Performance', 'Business', 'Biology', 'Education', 'Computer Science'])
plot
```
image

```python
plot['Business'].data is plot['Biology'].data  # True now :)
```

- https://github.com/holoviz/hvplot/pull/1350/commits/326ee64e06ebc721eed20dc92696611137cf25c8: HoloViews added in https://github.com/holoviz/holoviews/pull/6061 support for retaining `Pandas` indexes, which means that hvPlot no longer needs to call `reset_index` in a few cases, avoiding creating internal copies of the data. This change was made in https://github.com/holoviz/hvplot/pull/1350/commits/326ee64e06ebc721eed20dc92696611137cf25c8, by skipping some `reset_index` calls when the type of the dataset is a Pandas DataFrame (so far the Pandas Interface in HoloViews is the only one to have this support, so not Dask, not Geopandas, etc. which still require the `reset_index` calls). The preceding commits add various tests to make sure all the `reset_index` calls are at least called once in the unit tests. This commit https://github.com/holoviz/hvplot/pull/1350/commits/adbddf7f0c6890988da0fa390756c89a928e0d14 updates the instantiation of a `Dataset` object with an explicit list of kdims instead letting HoloViews infer them, as HoloViews when given a Pandas DataFrame doesn't include its indexes in the list of automatically inferred kdims.

- https://github.com/holoviz/hvplot/pull/1350/commits/abb485a0b253f3c8428f96af9af39b7d819db2ae: Only redim and/or relabel a HoloViews element when needed, as suggested in https://github.com/holoviz/hvplot/pull/1350#discussion_r1644152459 since apparently these operations return a clone even when not given any argument.

I hope these optimizations will help/fix this issue https://github.com/holoviz/hvplot/issues/501.

### CODE REVIEW

**Comment by droumis** [comment]
issue noted in meeting: "When printing the ndoverlay, the vdim is not a value label.. it's the label of the last element in the ndoverlay"

**Comment by philippjfr** [comment]
With https://github.com/holoviz/holoviews/pull/6262 this should now behave well.

**Comment by maximlt** [comment]
> When printing the ndoverlay, the vdim is not a value label.. it's the label of the last element in the ndoverlay

I confirm this is fixed, i.e. printing `out` from the code below:

```python
import numpy as np
import pandas as pd
import holoviews as hv
hv.extension('bokeh')

df = pd.DataFrame(np.random.random((5, 4)), columns=list('ABCD'))

value_label = 'value'
group_label = 'Variable'

charts = []
for column in df.columns:
    kdims = ['index']
    vdims = hv.Dimension(column, label=value_label)
    chart = hv.Curve(df, kdims, vdims)
    charts.append((column, chart))

out = hv.NdOverlay(charts, group_label, sort=False)
out
```

now displays:

```
:NdOverlay   [Variable]
   :Curve   [index]   (value)
```

File: hvplot/converter.py [hunk 1]
```diff
@@ -2123,8 +2123,9 @@ def chart(self, element, x, y, data=None):
 
         charts = []
         for c in y:
-            kdims, vdims = self._get_dimensions([x], [c])
-            chart = element(data, kdims, vdims).redim(**{c: self.value_label})
+            ydim = hv.Dimension(c, label=self.value_label)
+            kdims, vdims = self._get_dimensions([x], [ydim])
+            chart = element(data, kdims, vdims)
             charts.append((c, chart.relabel(**self._relabel).redim(**self._redim)))
```
Comment (philippjfr) [inline hvplot/converter.py:2129]: Maybe let's also skip `relabel` and `redim` if `self._relabel` and `self._redim` are empty respectively.

**Comment by droumis** [comment]
ready to merge?

**Comment by maximlt** [comment]
OP updated with a description of the changes made https://github.com/holoviz/hvplot/pull/1350#issue-2333794995.

**Review by philippjfr — APPROVED** [review]

**Comment by philippjfr** [comment]
Great work, all looked good to me and thanks for the new tests! Let's test this as much as possible, would love to see a new alpha release for that purpose.

**Pull request merged**